Support Non-Monotonic Saturated Oil FVF in PVTO

bska opened the pull request:
This PR relaxes the requirement that the saturated formation volume factor for oil ("Bo") be strictly increasing as a function of the dissolved gas/oil ratio ("Rs").  In a way, we partially revert commit dbf2a4c425af69804f552bab836ca1a3e838a2bd (PR #342).  We have seen cases for which the strictly increasing criterion fails to hold but which nevertheless run, at least if the monotonicity requirement only breaks for very high values of Rs.

We therefore introduce a special purpose system for detecting and reporting this situation.  In this case, output such as
```
Warning: Non-Monotonic Oil Formation Volume Factor Detected
  * PVTO [PVTNUM = 1]
  Record  9: FVF 1.23 at RS 123 is not greater than FVF 2.34 at RS 120
  Record 10: FVF 1.22 at RS 125 is not greater than FVF 1.23 at RS 123
  Record 11: FVF 1.21 at RS 125 is not greater than FVF 1.22 at RS 125
```
will be printed to the console and the `.PRT` file,  but the simulation will continue.

I have elected to make this a very narrow feature for the time being, but we can extend or rework this system if it proves useful.

bska commented:
jenkins build this please

bska commented:
I invite anyone who wishes to comment on the feature to do so.  Of particular interest is whether we even want this, if the proposed approach is reasonable, if the content of the diagnostic message is reasonable and if there are other/better ways to format the message.

joakim-hove commented:
>and if there are other/better ways to format the message.

Well - for the formatting I would certainly use the "about to actually work" fmtlib

joakim-hove commented:
> if the content of the diagnostic message is reasonable

The content seems fine to me - but I question whether we should output it all(?) - if we consider this something we can handle maybe we should just suck it up and do our job?

Problem is that we/our users do not really have an active approach to log levels - i.e. if there is a problem of sort I can ask for more logging, in the latter case this message should certainly be included.

joakim-hove commented on opm/parser/eclipse/EclipseState/Tables/PvtoTable.hpp:
```diff
@@ -17,22 +17,34 @@
+        struct FlippedFVF {
+            std::size_t i;
+            std::array<double, std::size_t{2}> Rs;
+            std::array<double, std::size_t{2}> Bo;
```
How about the slightly leaner: `std::array<double, 2> Rs;` - or even named members for `Rs[0]` and `Rs[1]`?

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Tables/TableManager.cpp:
```diff
@@ -1195,6 +1204,48 @@ namespace Opm {
+        for (const auto& pvto : tables) {
+            ++tableID; // One-based table ID
+            const auto& flipped = pvto.nonMonotonicSaturatedFVF();
+            if (flipped.empty()) { continue; }
```
This I just don't understand - it is so that even I would ask for a comment

OPMUSER commented:
I'm not sure if this is actual an error  or a warning in the commercial simulator, but is definitely suspect, so you should issue a warning message at the very least.
```
Warning: Non-Monotonic Oil Formation Volume Factor Detected
         PVTO [PVTNUM = 1]
         Record  9: FVF 1.23 at RS 123 is not greater than FVF 2.34 at RS 120
         Record 10: FVF 1.22 at RS 125 is not greater than FVF 1.23 at RS 123
         Record 11: FVF 1.21 at RS 125 is not greater than FVF 1.22 at RS 125
```

These type of things can cause computational efficiencies, so just because it runs does not mean that the results are correct.

plgbrts commented:
It's  a warning in the commercial simulator.

![image](https://user-images.githubusercontent.com/56635911/93592822-25641680-f9b3-11ea-872e-1a4c8a6459f4.png)

GitPaean commented:
I also suspect that this might cause convergence issues. A warning message might help when we need to fix problems relevant to this.

bska commented:
I have added a new commit which switches to using an `{fmt}` backend&mdash;implemented in terms of `fmt::format()`&mdash;to format the diagnostic message.  The new example diagnostic output is no longer indented and looks like
```
Warning: Non-Monotonic Oil Formation Volume Factor Detected
PVTO (/path/to/PVTO.INCL 6)
[PVTNUM = 1]
Record  9: FVF 1.234 at RS 123.456 is not greater than FVF 2.345 at RS 121.212
Record 10: FVF 1.233 at RS 123.457 is not greater than FVF 1.234 at RS 123.456
Record 11: FVF 1.233 at RS 123.460 is not greater than FVF 1.233 at RS 123.457
```
If #1939 is merged before this PR I will update this to use the `vector<string>` overload instead of formatting to a `std::ostringstream` before calling the `string` version of `OpmLog::warning()`.  I have not altered the `std::array` definitions.

The length of each `Record` line is 78 characters as of this change (assuming METRIC conventions, FIELD conventions not tested).  The length of the `PVTO` line depends on the number of characters in the full filepath.

bska commented:
jenkins build this please

alfbr commented:
@bska Thank you! Please consider revising to make it consistent with the output suggested in #1939 That is, like this:
```
Warning: Non-Monotonic Oil Formation Volume Factor in Keyword PVTO, PVTNUM=1
In /path/to/PVTO.INCL line 6
Record  9: FVF 1.234 at RS 123.456 is not greater than FVF 2.345 at RS 121.212
Record 10: FVF 1.233 at RS 123.457 is not greater than FVF 1.234 at RS 123.456
Record 11: FVF 1.233 at RS 123.460 is not greater than FVF 1.233 at RS 123.457
```
I like having all occurrences listed. I am a bit concerned that this may explode though (e..g, if every line in ZCORN is written out, it would be rather cumbersome). We should probably limit the amount of lines on terminal output to something practical, e.g., four records, and then end with a line like:

`See PRT-file for additional records, report limit reached.`

bska commented:
> ```
> Warning: Non-monotonic Oil Formation Volume Factor in keyword PVTO, PVTNUM=1
> In /path/to/PVTO.INCL line 6
> [...]
> ```

I added a separate line for the `PVTNUM` information to allow for the case of multiple tables (e.g., `PVTNUM = 1`, `PVTNUM = 4` and `PVTNUM = 7`) having these kinds of problems, in which case we'd have something along the lines of
```
Warning: Non-Monotonic Oil Formation Volume Factor Detected
PVTO (/path/to/PVTO.INCL 6)
[PVTNUM = 1]
Record  9: FVF 1.234 at RS 123.456 is not greater than FVF 2.345 at RS 121.212
Record 10: FVF 1.233 at RS 123.457 is not greater than FVF 1.234 at RS 123.456
Record 11: FVF 1.233 at RS 123.460 is not greater than FVF 1.233 at RS 123.457
[PVTNUM = 4]
Record 32: FVF 1.234 at RS 321.098 is not greater than FVF 2.345 at RS 320.987
[PVTNUM = 7]
Record 5: FVF 1.234 at RS 135.791 is not greater than FVF 1.357 at RS 130.000
Record 8: FVF 1.233 at RS 150.105 is not greater than FVF 1.234 at RS 145.987
```
If you don't think that's a likely scenario then I'll rewrite it in your suggested form.  File/line information on a single line is a good idea and I'll do that regardless.

> We should probably limit the amount of lines on terminal output to something practical, e.g., four records, and then end with a line like:

I like that idea.  I don't know how easy that's to do in our current logging system (that system is mostly based on the assumption that everything goes to both the `.PRT` and the console), but I'll look into it.

> added one additional space to allow three digit record number.

Just for the record, the field width for the record number is [adjusted dynamically](https://github.com/OPM/opm-common/blob/ca2cbd0e69fe547a5d010c77047b5af66cc653af/src/opm/parser/eclipse/EclipseState/Tables/TableManager.cpp#L1239) based on the [maximum record ID](https://github.com/OPM/opm-common/blob/ca2cbd0e69fe547a5d010c77047b5af66cc653af/src/opm/parser/eclipse/EclipseState/Tables/TableManager.cpp#L1265).

alfbr commented:
> If you don't think that's a likely scenario then I'll rewrite it in your suggested form.

Actually, I would prefer a new Warning message emitted if it occures in another PVTNUM. There are usually very few PVT regions.

>File/line information on a single line is a good idea and I'll do that regardless.

Thank you!

> I like that idea. I don't know how easy that's to do in our current logging system (that system is mostly based on the assumption that everything goes to both the `.PRT` and the console), but I'll look into it.

It is already done for a selection of warnings, the output keywords comes to mind, where a maximum of eight warnings are emitted to the terminal. If my memory serves me, all are found in the PRT file though. I believe it is similar for some of the table checking, but this is just from the top of my head. I will need to double check to give you precise information.

> Just for the record, the field width for the record number is [adjusted dynamically]

Ah, that is elegant. Then please excuse my ignorance.

bska commented:
jenkins build this please

bska commented:
@alfbr : I have pushed a new update which implements your earlier [suggestion](https://github.com/OPM/opm-common/pull/1938#issuecomment-696267078). We now output at most four (4) records of flipped oil FVF values to the console, but output all such records to the `.PRT` file.

New example output (with console record limit set to 1 for expository purposes):

### To the Console (Usually in Colour) ###
```
Warning: Non-Monotonic Oil Formation Volume Factor Detected in Keyword PVTO, PVTNUM=1
In /path/to/PVTO.INCL line 6
Record  9: FVF 1.234 at RS 123.456 is not greater than FVF 2.345 at RS 121.212
Report limit reached, see PRT-file for additional records.
```

### To the PRT File ###
```
Warning: Non-Monotonic Oil Formation Volume Factor Detected in Keyword PVTO, PVTNUM=1
In /path/to/PVTO.INCL line 6
Record  9: FVF 1.234 at RS 123.456 is not greater than FVF 2.345 at RS 121.212
Record 10: FVF 1.233 at RS 123.457 is not greater than FVF 1.234 at RS 123.456
Record 11: FVF 1.233 at RS 123.460 is not greater than FVF 1.233 at RS 123.457
```

alfbr commented:
Looks great! My apologies for being unclear though, I believe we could easily to four lines of records, or three is sufficient I guess, then it would be like this:

```
Warning: Non-Monotonic Oil Formation Volume Factor Detected in Keyword PVTO, PVTNUM=1
In /path/to/PVTO.INCL line 6
Record  9: FVF 1.234 at RS 123.456 is not greater than FVF 2.345 at RS 121.212
Record 10: FVF 1.233 at RS 123.457 is not greater than FVF 1.234 at RS 123.456
Record 11: FVF 1.233 at RS 123.460 is not greater than FVF 1.233 at RS 123.457
Report limit reached, see PRT-file for additional records.
```
While PRT file would look like this (assuming four records in violation):
```
Warning: Non-Monotonic Oil Formation Volume Factor Detected in Keyword PVTO, PVTNUM=1
In /path/to/PVTO.INCL line 6
Record  9: FVF 1.234 at RS 123.456 is not greater than FVF 2.345 at RS 121.212
Record 10: FVF 1.233 at RS 123.457 is not greater than FVF 1.234 at RS 123.456
Record 11: FVF 1.233 at RS 123.460 is not greater than FVF 1.233 at RS 123.457
Record 12: FVF 1.233 at RS 123.470 is not greater than FVF 1.231 at RS 123.458
```

bska commented:
> I believe we could easily do four lines of records, or three is sufficient I guess, then it would be like this:

Okay, so if the record limit is `N` (= 4 in your example; although `N > 0` for this to make sense), then we always output at most `N` record lines to the console, while the `.PRT` always gets all flipped records.  That means that

1. If the number of flipped records is less than or equal to `N`, then we output all of them to the console

2. If the number of flipped records is greater than `N`, then we output the `N - 1` first flipped records and then one line saying "Report limit reached..."

Is that what you had in mind?

alfbr commented:
That is better than what I had in mind. Please go for it with N=4.

atgeirr commented:
Sorry for perhaps just adding noise, but is it not possible to use the existing message tag system to limit the messages?

alfbr commented:
> Sorry for perhaps just adding noise, but is it not possible to use the existing message tag system to limit the messages?

It is not noise, it is a valid point. We need to unify the presentation of messages, so using common system components where applicable is a natural step.

bska commented:
> is it not possible to use the existing message tag system to limit the messages?

I don't see how, but I haven't studied the limiter closely.  This PR creates a single warning message that happens to span multiple lines, but as far as I can tell the tag system is about limiting the total number of warnings.  Perhaps I'm just missing something essential.

alfbr commented:
> > is it not possible to use the existing message tag system to limit the messages?
> 
> I don't see how, but I haven't studied the limiter closely. This PR creates a single warning message that happens to span multiple lines, but as far as I can tell the tag system is about limiting the total number of warnings. Perhaps I'm just missing something essential.

Then noise it was. Worth checking out though, thanks!

bska commented:
@alfbr : I've pushed a final update which implements the approach [outlined earlier](https://github.com/OPM/opm-common/pull/1938#issuecomment-696714726).  I've checked that it does what I want to on a real case, admittedly by artificially reducing the limit to `2` instead of `4` (the case only has `3` flipped records ).  The update uses `4` as agreed earlier.

I now consider this PR to be functionally complete and I'd like to mark it ready for review.

The synthetic example output from before now looks like

### To the Console (Usually in Colour) ###
```
Warning: Non-Monotonic Oil Formation Volume Factor Detected in Keyword PVTO, PVTNUM=1
In /path/to/PVTO.INCL line 6
Record  9: FVF 1.234 at RS 123.456 is not greater than FVF 2.345 at RS 121.212
Record 10: FVF 1.233 at RS 123.457 is not greater than FVF 1.234 at RS 123.456
Record 11: FVF 1.233 at RS 123.460 is not greater than FVF 1.233 at RS 123.457
Report limit reached, see PRT-file for additional records.
```

### In the PRT File ###
```
Warning: Non-Monotonic Oil Formation Volume Factor Detected in Keyword PVTO, PVTNUM=1
In /path/to/PVTO.INCL line 6
Record  9: FVF 1.234 at RS 123.456 is not greater than FVF 2.345 at RS 121.212
Record 10: FVF 1.233 at RS 123.457 is not greater than FVF 1.234 at RS 123.456
Record 11: FVF 1.233 at RS 123.460 is not greater than FVF 1.233 at RS 123.457
Record 12: FVF 1.233 at RS 123.461 is not greater than FVF 1.233 at RS 123.460
Record 13: FVF 1.233 at RS 123.470 is not greater than FVF 1.233 at RS 123.461
```
---

> > > is it not possible to use the existing message tag system to limit the messages?
> > 
> > I don't see how, but I haven't studied the limiter closely. This PR creates a single warning message that happens to span multiple lines, but as far as I can tell the tag system is about limiting the total number of warnings. Perhaps I'm just missing something essential.
> 
> Worth checking out though, thanks!

As an aside, if we get #1939 into master then it will be feasible to add a general system for limiting the number of lines in a single message.  In that case, I suggest we generalise and move this PR's ad-hoc [limiting logic](https://github.com/OPM/opm-common/blob/f4331a2513324e41cae2c9bb70f1f3f7a2fa01e7/src/opm/parser/eclipse/EclipseState/Tables/TableManager.cpp#L1303-L1308) to the message limiting system.

alfbr commented:
From my end this is perfect, thank you!

bska commented:
>> I now consider this PR to be functionally complete and I'd like to mark it ready for review.
>
> From my end this is perfect, thank you!

Thanks.  Then I'll mark it ready for review.

I also looked into the possibility of including line information (file is implied) for each individual record, but it appears we don't actually track that information for constituent records of a keyword.  Maybe that's something we could consider doing, at least for a select few table-related keywords.  I have no idea how hard it would be to do so however., so this is just speculation for now.

bska commented:
jenkins build this please

bska commented:
jenkins build this please

alfbr commented:
This is perfect from my end, great if we can get it reviewed/merged.

bska commented:
jenkins build this please

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Tables/TableManager.cpp:
```diff
@@ -1195,6 +1208,114 @@ namespace Opm {
+                console << record << '\n';
+            }
+            else if (recordIx + 1 == consoleRecordLimit) {
+                console << "Report limit reached, see PRT-file for additional records.\n";
```
Is there a reason for not using the message limiter functionality which is already in OpmLog?

atgeirr replied:
I will try to answer this, to see if I understood it correctly: The whole part here just prints a single message to each of the logs, that can have many lines (one per record). The code here limits the lines in the console version of the message. The message limits would only apply if each record was a separate message.

@bska please shoot this down if wrong...

bska replied:
The limiter addresses the number of messages, not the number of lines per individual message.

bska replied:
> I will try to answer this, to see if I understood it correctly

@atgeirr: Yes, that is correct.

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Tables/TableManager.cpp:
```diff
@@ -1195,6 +1208,114 @@ namespace Opm {
+        // Print file gets all flipped records.
+        const auto numDigitsRecordID  = nDigit(flipped_Bo.back().i + 1);
+        const auto numRecords         = flipped_Bo.size();
+        const auto consoleRecordLimit = decltype(numRecords){4};
```
Using `decltype(numRecords){4};` instead of `4;` reduces readability for no gain in my opinion.

bska replied:
It avoid "signed vs. unsigned" warnings.

atgeirr replied:
I agree. Inside very generic library code it can pay off to be so meticulous, but I doubt its value in this context.

atgeirr replied:
The messages here got misleading order... I think applying the "const auto var = " pattern here is not necessary. Better with `std::size_t consoleRecordLimit = 4;`?

bska replied:
> Better with `std::size_t consoleRecordLimit = 4;`?

Except we don't know that `std::size_t` is the correct type.  If you want to actually be correct, you need to spell it out as
```C++
const std::vector<PvtoTable::FlippedFVF>::size_type consoleRecordLimit = 4;
```
and at that point the name of the actual object (`consoleRecordLimit`) is pushed very far towards the right-hand margin.

bska replied:
> Better with `std::size_t consoleRecordLimit = 4;`?

I suppose I could force all counters to `int` by writing it as
```C++
const int  numDigitsRecordID  = nDigit(flipped_Bo.back().i + 1);
const int  numRecords         = static_cast<int>(flipped_Bo.size());
const int  consoleRecordLimit = 4;
const int  floatPrecision     = 3;
const bool consoleWriteAll    = numRecords <= consoleRecordLimit;
```
but there's going to be an explicit conversion **somewhere** in that block.

atgeirr replied:
We should write code that Joe Freshman Coder would understand.

So unless necessary, we avoid decltype(). Using that tells the reader: "this is advanced stuff", and that is not true here. Also, std::size_t is *a* correct type to use here, the full size_type specification is over-engineering it, reduces readability, and, again, tells the reader "this is advanced".

About forcing counters to `int`: I used to advocate that, but I have for a long time acknowledged that battle is lost, so whenever we need to compare integer types, follow the path of least resistance. I do not know from the code snippet what parts are unsigned or not. Outside very generic utility libraries I do not consider that a benefit, but a drawback. It is fine to use auto, but use a specific type when necessary.

Would not the simplest solution avoiding a warning be `const auto consoleRecordLimit = 4u;`?

bska replied:
> the full `size_type` specification is over-engineering it, reduces readability, and, again, tells the reader "this is advanced".

Which is why `auto` is so much better for this.  It tells the reader, "I don't really care what the particular type of this object is, just make it the same as that initializer over there".

In any case, I've pushed an update that forces the counters to `std::size_t`.  This is preferable to using `auto = 4u` because that will generate `unsigned int` which is definitely wrong on many 64 bit platforms.

bska commented:
jenkins build this please

joakim-hove commented:
Thank you for the updates!
